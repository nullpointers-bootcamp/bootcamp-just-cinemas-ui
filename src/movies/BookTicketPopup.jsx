import React from "react";
import Popup from "reactjs-popup";
import PropTypes from "prop-types";
import moment from "moment";
import "./bookingTicketPopup.css";
import cx from "classnames";
import ShowInformation from "./ShowInformation";

class BookTicketPopup extends React.Component {
  componentDidMount() {
    this.props.fetchShowInformation(
      moment().format("YYYY-MM-DD"),
      this.props.movieId
    );
  }

  fetchShowInformation = day => {
    this.props.fetchShowInformation(day, this.props.movieId);
  };

  renderDate(day, index) {
    return (
      <div
        key={index}
        className={cx("booking-date-item", {
          "date-selected": this.props.bookingDate === day
        })}
        onClick={() => this.fetchShowInformation(day)}
      >
        <div>{moment(day).format("ddd")}</div>
        <div>{moment(day).format("DD")}</div>
      </div>
    );
  }

  renderDates() {
    const days = this.getNextNDaysFromGivenDate(Date.now(), 7);
    return (
      <div className="booking-dates-wrapper">
        <div className="booking-date-item booking-date-label">Date :</div>
        {days.map((day, index) => this.renderDate(day, index))}
      </div>
    );
  }

  getNextNDaysFromGivenDate(date, nDays) {
    let days = [];
    for (let i = 0; i < nDays; i++) {
      const today = moment(date);
      days.push(
        moment(today)
          .add(i, "days")
          .format("YYYY-MM-DD")
      );
    }
    return days;
  }

  render() {
    const { show, onClose, bookingDate, showInformation } = this.props;
    return (
      <Popup open={show} onClose={onClose}>
        <div>
          <div className="book-ticket-wrapper">{this.renderDates()}</div>
          {bookingDate && showInformation.length ? (
            <ShowInformation shows={showInformation} />
          ) : null}
        </div>
      </Popup>
    );
  }
}

BookTicketPopup.defaultProps = {
  show: false,
  movieId: null,
  bookingDate: null,
  showInformation: []
};

BookTicketPopup.propTypes = {
  show: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  fetchShowInformation: PropTypes.func.isRequired,
  movieId: PropTypes.string.isRequired,
  bookingDate: PropTypes.string.isRequired,
  showInformation: PropTypes.array.isRequired
};
export default BookTicketPopup;
