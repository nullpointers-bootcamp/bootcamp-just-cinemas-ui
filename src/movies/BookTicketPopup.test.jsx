import React from "react";
import renderer from "react-test-renderer";
import BookTicketPopup from "./BookTicketPopup";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

it("should show the movie ticket pop ", () => {
  const tree = renderer
    .create(<BookTicketPopup show={true} onClose={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("should hide the movie ticket pop ", () => {
  const tree = renderer
    .create(<BookTicketPopup show={false} onClose={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("should get next 5 days from current date ", () => {
  const wrapper = shallow(<BookTicketPopup show={false} onClose={() => {}} />);
  expect(wrapper.instance().getNextNDaysFromGivenDate("2019-02-20", 5)).toEqual(
    ["2019-02-20", "2019-02-21", "2019-02-22", "2019-02-23", "2019-02-24"]
  );
});
