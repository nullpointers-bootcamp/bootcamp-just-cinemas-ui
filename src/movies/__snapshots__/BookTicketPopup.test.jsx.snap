// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should hide the movie ticket pop  1`] = `null`;

exports[`should highlight date when it is selected  1`] = `
Array [
  <div
    style={
        Object {
            "left": "0px",
            "position": "absolute",
            "top": "0px",
          }
    }
/>,
  <div
    className="popup-overlay"
    onClick={[Function]}
    style={
        Object {
            "background": "rgba(0, 0, 0,0.5)",
            "bottom": "0",
            "display": "flex",
            "left": "0",
            "position": "fixed",
            "right": "0",
            "top": "0",
            "zIndex": "999",
          }
    }
>
    <div
        className="popup-content "
        onClick={[Function]}
        style={
            Object {
                "background": "rgb(255, 255, 255)",
                "border": "1px solid rgb(187, 187, 187)",
                "margin": "auto",
                "padding": "5px",
                "position": "relative",
                "width": "50%",
              }
        }
    >
        <div>
            <div
                className="book-ticket-wrapper"
            >
                <div
                    className="booking-dates-wrapper"
                >
                    <div
                        className="booking-date-item booking-date-label"
                    >
                        Date :
                    </div>
                    <div
                        className="booking-date-item date-selected"
                        onClick={[Function]}
                    >
                        <div>
                            Wed
                        </div>
                        <div>
                            20
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Thu
                        </div>
                        <div>
                            21
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Fri
                        </div>
                        <div>
                            22
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Sat
                        </div>
                        <div>
                            23
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Sun
                        </div>
                        <div>
                            24
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Mon
                        </div>
                        <div>
                            25
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Tue
                        </div>
                        <div>
                            26
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>,
]
`;

exports[`should show the movie ticket pop  1`] = `
Array [
  <div
    style={
        Object {
            "left": "0px",
            "position": "absolute",
            "top": "0px",
          }
    }
/>,
  <div
    className="popup-overlay"
    onClick={[Function]}
    style={
        Object {
            "background": "rgba(0, 0, 0,0.5)",
            "bottom": "0",
            "display": "flex",
            "left": "0",
            "position": "fixed",
            "right": "0",
            "top": "0",
            "zIndex": "999",
          }
    }
>
    <div
        className="popup-content "
        onClick={[Function]}
        style={
            Object {
                "background": "rgb(255, 255, 255)",
                "border": "1px solid rgb(187, 187, 187)",
                "margin": "auto",
                "padding": "5px",
                "position": "relative",
                "width": "50%",
              }
        }
    >
        <div>
            <div
                className="book-ticket-wrapper"
            >
                <div
                    className="booking-dates-wrapper"
                >
                    <div
                        className="booking-date-item booking-date-label"
                    >
                        Date :
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Wed
                        </div>
                        <div>
                            20
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Thu
                        </div>
                        <div>
                            21
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Fri
                        </div>
                        <div>
                            22
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Sat
                        </div>
                        <div>
                            23
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Sun
                        </div>
                        <div>
                            24
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Mon
                        </div>
                        <div>
                            25
                        </div>
                    </div>
                    <div
                        className="booking-date-item"
                        onClick={[Function]}
                    >
                        <div>
                            Tue
                        </div>
                        <div>
                            26
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>,
]
`;
